crates/test-programs/wasi-tests/src/bin/interesting_paths.rs

```diff
@@ -1,152 +1,106 @@
     // Create a directory in the scratch directory.
-    create_dir(dir_fd, "dir");
+    wasi::path_create_directory(dir_fd, "dir").expect("creating dir");
 
     // Create a directory in the directory we just created.
-    create_dir(dir_fd, "dir/nested");
+    wasi::path_create_directory(dir_fd, "dir/nested").expect("creating a nested dir");
 
     // Create a file in the nested directory.
     create_file(dir_fd, "dir/nested/file");
 
     // Now open it with an absolute path.
-    let mut file_fd: wasi_unstable::Fd = wasi_unstable::Fd::max_value() - 1;
-    let mut status = wasi_path_open(dir_fd, 0, "/dir/nested/file", 0, 0, 0, 0, &mut file_fd);
     assert_eq!(
-        status,
-        wasi_unstable::raw::__WASI_ENOTCAPABLE,
-        "opening a file with an absolute path"
-    );
-    assert_eq!(
-        file_fd,
-        wasi_unstable::Fd::max_value(),
-        "failed open should set the file descriptor to -1",
+        wasi::path_open(dir_fd, 0, "/dir/nested/file", 0, 0, 0, 0)
+            .expect_err("opening a file with an absolute path")
+            .raw_error(),
+        wasi::ERRNO_NOTCAPABLE,
+        "errno should be ERRNO_NOTCAPABLE"
     );
 
     // Now open it with a path containing "..".
-    status = wasi_path_open(
+    let mut file_fd = wasi::path_open(
         dir_fd,
         0,
         "dir/.//nested/../../dir/nested/../nested///./file",
         0,
         0,
         0,
         0,
-        &mut file_fd,
-    );
-    assert_eq!(
-        status,
-        wasi_unstable::raw::__WASI_ESUCCESS,
-        "opening a file with \"..\" in the path"
-    );
+    )
+    .expect("opening a file with \"..\" in the path");
     assert_gt!(
         file_fd,
-        libc::STDERR_FILENO as wasi_unstable::Fd,
+        libc::STDERR_FILENO as wasi::Fd,
         "file descriptor range check",
     );
-    close_fd(file_fd);
+    wasi::fd_close(file_fd).expect("closing a file");
 
     // Now open it with a trailing NUL.
-    status = wasi_path_open(dir_fd, 0, "dir/nested/file\0", 0, 0, 0, 0, &mut file_fd);
     assert_eq!(
-        status,
-        wasi_unstable::raw::__WASI_EILSEQ,
-        "opening a file with a trailing NUL"
-    );
-    assert_eq!(
-        file_fd,
-        wasi_unstable::Fd::max_value(),
-        "failed open should set the file descriptor to -1",
+        wasi::path_open(dir_fd, 0, "dir/nested/file\0", 0, 0, 0, 0)
+            .expect_err("opening a file with a trailing NUL")
+            .raw_error(),
+        wasi::ERRNO_ILSEQ,
+        "errno should be ERRNO_ILSEQ",
     );
 
     // Now open it with a trailing slash.
-    status = wasi_path_open(dir_fd, 0, "dir/nested/file/", 0, 0, 0, 0, &mut file_fd);
-    assert_eq!(
-        status,
-        wasi_unstable::raw::__WASI_ENOTDIR,
-        "opening a file with a trailing slash"
-    );
     assert_eq!(
-        file_fd,
-        wasi_unstable::Fd::max_value(),
-        "failed open should set the file descriptor to -1",
```

Comment:
One unfortunate thing about this change is that it means we no longer test the value of `file_fd` coming out of the WASI syscall in the failure case. The `wasi` crate is right to hide this detail, but it is an observable aspect of WASI if one codes at the raw WASI level.

I'm undecided on what we should do here. On one hand, I agree that using the higher-level interfaces makes these tests much nicer. On the other, we will eventually want detailed low-level test coverage at the raw WASI level.

I wonder if it would make sense to do something in the `wasi` crate? We could add asserts for the returned values on the error paths, maybe controlled by a cargo feature so that they'd be off by default but we could enable them here. Thoughts?

Reply:
Yeah, I wasn't sure how to deal with it either. I really like the way the tests look now, as thanks to the latest version of the `wasi` crate, they are much cleaner now. Having said that, we are now not dealing directly with the syscalls but rather with their higher-level wrappers.

Personally, I like your idea to perform the asserts in the `wasi` crate and have them feature-gated so that they are off by default. This would probably mean we could move the range check assert

```rust
assert_gt!(
    fd,
    libc::STDERR_FILENO as wasi::Fd,
    "file descriptor range check",
);
```

to the `wasi` crate as well. So it's :+1: from me.